### PR TITLE
refactor(*): remove gt and ge

### PR DESCRIPTION
**Pull request by gebner**
This changes `>` and `≥` to be notations instead of definitions.  The only downside is that you can't use them in calc-blocks anymore (the fix is reverse the whole calc-block so that it uses `<` or `≤`, resp.).

### CODE REVIEW

**Comment by cipher1024** [comment]
Why aren't they usable in `calc` blocks anymore? If we used `abbreviation` instead of `notation`, would that be good enough?

**Comment by jcommelin** [comment]
I would also very much like to keep them available in calc-blocks.
On the up side, does this mean that the `>`-linter can now be removed? Because `simp` will see through the notation, right?

File: library/init/algebra/order.lean [hunk 1]
```diff
@@ -54,7 +54,7 @@ lemma le_of_eq [preorder α] {a b : α} : a = b → a ≤ b :=
 lemma le_antisymm_iff [partial_order α] {a b : α} : a = b ↔ a ≤ b ∧ b ≤ a :=
 ⟨λe, ⟨le_of_eq e, le_of_eq e.symm⟩, λ⟨h1, h2⟩, le_antisymm h1 h2⟩
 
-@[trans] lemma ge_trans [preorder α] : ∀ {a b c : α}, a ≥ b → b ≥ c → a ≥ c :=
+lemma ge_trans [preorder α] : ∀ {a b c : α}, a ≥ b → b ≥ c → a ≥ c :=
```
Comment (jcommelin) [inline library/init/algebra/order.lean]: Why are you removing this attribute? Doesn't that mean that
```lean
  transitivity c,
```
will no longer work when I'm proving `a ≥ b`?

Comment (digama0) [reply library/init/algebra/order.lean]: Now that `≥` is a notation, `ge_trans` is identical to ` le_trans` (except the two arguments are in the opposite order). When you are proving `a ≥ b`, lean will see that you are actually proving `b ≤ a`, and so `transitivity c` will work, and will apply `le_trans` instead, meaning that the two subgoals will be in the opposite order from today.

Comment (jcommelin) [reply library/init/algebra/order.lean]: Aha, fair enough.

**Comment by gebner** [comment]
> Why aren't they usable in calc blocks anymore?

The immediate reason is because the calc parser joins the proofs in the calc proof using transitivity steps.  So consider the following example:
```lean
calc a = b    : p1
   ... > c    : p2
```
Here `calc` needs to combine `a = b` and `dummy > c` via transitivity.  However, the second expression is actually `c < dummy` and Lean doesn't know how to apply transitivity here.  (The calc parser also gets confused since dummy is now the second argument.)

> If we used abbreviation instead of notation, would that be good enough?

The definitions `ge` and `gt` are already reducible.  AFAICT the only difference between `abbreviation` and `@[reducible, inline] def` is that that the unifier unfolds the abbreviation a bit more eagerly.

I'm not sure what goes wrong exactly, but I think it boils down to the fact that `e.get_app_fn.const_name` returns different names for `a < b` and `a > b`.  Turning `ge` and `gt` into an abbreviation doesn't change that.

@digama0 @fpvandoorn IIRC you've been heavily involved in the war against `>`.  Do you have a small MWE of what goes wrong with `>`?

**Comment by gebner** [comment]
>  On the up side, does this mean that the >-linter can now be removed? Because simp will see through the notation, right?

Yes!  That's the motivation for this change.

**Comment by Kha** [comment]
> I'm not sure what goes wrong exactly, but I think it boils down to the fact that `e.get_app_fn.const_name` returns different names for `a < b` and `a > b`. Turning `ge` and `gt` into an abbreviation doesn't change that.

The big issue is head matching as used by `simp` and others, right? There is the `add_key_equivalence` command [that is supposed to solve this issue](https://github.com/leanprover/lean/blob/master/src/library/tactic/kabstract.h#L18), but AFAICS it's never been tried in either corelib or mathlib...

**Comment by gebner** [comment]
> The big issue is head matching as used by simp and others, right?

I assume so, but I'm not an expert on the gt-issues.

> There is the add_key_equivalence command that is supposed to solve this issue.

Unfortunately, the `add_key_equivalence` only affects `kabstract` used by `rw`.  The simplifier uses the `head_index` class, which literally computes `const_name(get_app_fn(lhs))`: https://github.com/leanprover-community/lean/blob/ec1613aef1eee72e601f192b16740629c6d49690/src/library/head_map.cpp#L23

**Comment by Kha** [comment]
Ah, you're right.

**Comment by ChrisHughes24** [comment]
Maybe this is for another PR, but should we do the same for `ne` as well?

**Comment by digama0** [comment]
I don't think so. `ne a b` is actually shorter than `not (eq a b)`, and also there are theorems in the `ne` namespace like `ne.symm` that would be lost if it was turned into a notation. Plus `ne` is in simp normal form so there isn't a need to eliminate it.

**Comment by cipher1024** [comment]
If we could add the right keys in the right tables rather than removing `>`, I'd prefer that. And if we could have a notion of synonym that can be reused, that would be even better. I can think of the same treatment for `>>=` vs `=<<`. Maybe we need an attribute?

**Comment by rwbarton** [comment]
One mild reason for `>` to be a definition rather than notation is that it means the pretty-printer can reproduce what the user wrote, rather than always producing `<` (or worse, always producing `>`).

**Pull request closed without merging**